### ISSUE TITLE
TwelveMonkeys Java 9+ issues

### ISSUE THREAD

**Issue by uttamburman** [haraldk/TwelveMonkeys#507]
Hi
I tried using library with Java 8 and it works flawless.
When I attempt to run it on Java 9 or Java 11 versions, java imageio plugins (or readers) get picked which do not work for files which were earlier readable on java 8.

I tried using SPI and setting TIFFImageReaderSpi and run it. it throws error - 
javax.imageio.IIOException: Invalid TIFF byte order mark 'II' or 'MM'

Reading in to one of the issues, I found that it may not be TIFF file , but it is as IIORegistry metadata says and twelvemonkeys read it when over java 8. 

**1) Just trying to figure out where am I doing wrong?** 
 Code is something like-
IIORegistry registry = IIORegistry.getDefaultInstance();
ImageReaderSpi readerSPI = registry.getServiceProviderByClass(TIFFImageReaderSpi.class);
com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader reader = new com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader(SPI);

Let me know, how it can be resolved.

And Yes, to make it work I had to change the constructor from default to public in 3.5-SNAPSHOT -
public TIFFImageReader(final ImageReaderSpi provider) {
        super(provider);
    }

as it wasn't accessible. 

**2) Is it wrong to do that as it is public in imageIO java 11.**

**Comment by uttamburman** [comment]
Tried using same code with java 8 . It fails at TIFFReader:100 where it identifies ByteOrder when using with SPI, but passes when plugin is scanned from imageio.

**Comment by Schmidor** [comment]
The ImageReaderSpi ist not designed to be used directly.
I'm using a loop and instanceof checks to prioritize TwelveMonkeys over other readers:
```
public static ImageReader getReader( ImageInputStream input ) {
    Iterator<ImageReader> readers = ImageIO.getImageReaders( input );
    ImageReader reader = null;
    while (readers.hasNext()) {
      reader = readers.next();
      if (reader instanceof ImageReaderBase) {
        break;
      }
    }
    return reader;
  }
```

**Comment by uttamburman** [comment]
I got it solved without using SPI.

Tried checking readers in ImageReader(Iterator) , it gave 2 readers in Java 11 and 1 in Java 8. 
Placed a check to use twelvemonkeys' (vendor)instead of ImageIO. It works fine.

Code-

`Iterator<ImageReader> imageReaders = ImageIO.getImageReaders(imageStream);
while (imageReaders.hasNext()) {
				ImageReader reader = (ImageReader) imageReaders.next();

				if (!reader.toString().contains("twelvemonkeys")) {
					System.out.println("Found some other Reader!");
					continue;
				}
//...... Some other stuff!
}
`

**Comment by haraldk** [comment]
Hi,

Sorry for slow response, have been out of office for a few days. Thanks @Schmidor for stepping in! :-)

I believe the issue you see is that Java from version 9 and onward bundle its own TIFF plugin (see [JEP 262 for details](https://openjdk.java.net/jeps/262)), which is selected before our plugin. I will fix this in an upcoming 4.0 release, but unfortunately, we're not there yet. In the meantime, you can select the reader as Oliver suggests, or you can explicitly set the ordering or uninstall the bundled plugin [as shown in this SO anwer](https://stackoverflow.com/a/38198991/1428606) (just change the plugin names). This latter approach will make `ImageIO.read(...)` select the TwelveMonkeys plugin, without the extra code at the call site.

If you still think there is something wrong with the library (detecting/reading TIFF), please attach the file you are trying to read, and a minimal code sample to reproduce. If not, I'll close this issue.

PS: If you insist on using the SPI directly (not recommended), the "correct" way to create a reader is:

    ImageReaderSpi spi = ...;
    ImageReader reader = spi.createReaderInstance();

No need to make the constructor public. :-)


Best regards,

-- 
Harald K

**Comment by uttamburman** [comment]
Yes, that worked for me. Thanks!
